### ISSUE TITLE
Composed TypedDicts don't work with `.update()`

### ISSUE THREAD

**Issue by AaronAsAChimp** [microsoft/pyright#10922]
When using composed TypedDicts the `.update()` method fails to pass type checking when adding properties specified by one of its parent classes. Upcasting the composite dict allows the update call to type check properly.

**Code**

```python
from typing import TypedDict, cast

class Mixin(TypedDict):
    a_prop: str
    
class Composed(Mixin):
    b_prop: str
    
composite: Composed = {
    'b_prop': '',
    'a_prop': ''
}

mixin: Mixin = {
    'a_prop': ''
}

cast(Mixin, composite).update(mixin)

composite.update(mixin)
```

Error:

```
No overloads for "update" match the provided arguments

Argument of type "Mixin" cannot be assigned to parameter "__m" of type "Partial[Composed]" in function "update"
  "b_prop" is an incompatible type
    "object" is not assignable to "str"
```

**Comment by erictraut** [comment]
Pyright's behavior here is correct and in compliance with the typing spec. TypedDicts are structural types, and they are "open" by default. That means a value that conforms with type `Mixin` can have additional keys beyond `a_prop` — including a `b_prop` key. These additional keys can be any `object` type, so it's possible for a `MixIn` to have a `b_prop` key with a value that's something other than a `str`.

To make this type safe, you'll need to leverage the new functionality provided by the recently-approved [PEP 728](https://peps.python.org/pep-0728/).

**Comment by karolyi** [comment]
I was wondering why going through with `.items()` on a TypedDict only yielded `object` as values. It seems this is the explanation for it.

My problem is, no cpython version recognizes the `closed=True` parameter. I've tried with 3.11-3.13, to no avail:
```python
Python 3.13.7 (main, Aug 15 2025, 12:34:02) [GCC 15.2.1 20250813] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import TYPE_CHECKING, NotRequired, TypedDict
>>> class SomeTypedDict(TypedDict, closed=True):
...     prefix: str
...     
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    class SomeTypedDict(TypedDict, closed=True):
        prefix: str
TypeError: _TypedDictMeta.__new__() got an unexpected keyword argument 'closed'
>>> 
```

Am I missing something here?

**Comment by karolyi** [comment]
Hm, it seems that this is not a runtime option.

So as it seems now, if one wants to have this handled properly with typing, all the classes that use this will have to be defined twice, once with `TYPE_CHECKING` and once without, e.g.:

```python
from typing import TYPE_CHECKING, TypedDict

if TYPE_CHECKING:
    class SomeTypedDict(TypedDict, closed=True):
        prefix: str
else:
    class SomeTypedDict(TypedDict):
        prefix: str
```

Doing this for each classes, just to satisfy pyright is crazy.

**Comment by erictraut** [comment]
> no cpython version recognizes the `closed=True` parameter

Support for this will be in Python 3.14. If you need it to work in older versions of Python, you can use `typing_extensions`, which exports a `TypedDict` class that accepts `closed=True`.

**Comment by karolyi** [comment]
Hey Eric,

thanks for the responding even to a closed issue. I'll be looking into that. I'm already using `typing_extensions`, I just thought loading it would patch TypedDict to automatically support this.

Will test tomorrow, thanks again.
